### ISSUE TITLE
captain.log - IPFS Node.js implementation

### ISSUE THREAD

**Issue by diasdavid** [ipfs/node-ipfs#30]
Aye mates!

This is your captain speaking, the IPFS Node.js ship as been to set sail and good winds are blowing! Thank you for all the messages and enthusiasm shared so far, we truly appreciate your support and since some of you are eager to get their hands on code and help with the development, I'll keep this issue updated every week with what has been going on and where you can help the most. To make sure you don't miss any updates, click the 'watch' button on the repo.

### LOG - Week of 21st of September 

**Plan:**

- Finish Swarm revisit 
  - https://github.com/diasdavid/node-ipfs-swarm/issues/8
  - https://github.com/diasdavid/node-ipfs-swarm/tree/revisit
- Implement node-libp2p bindings for every other module that is part of libp2p and have a alpha release people can experiment with
  - https://github.com/diasdavid/node-libp2p

**Where you can help during this week:**

- If crypto is your thing, take a look at 
  - https://github.com/diasdavid/webcrypto/issues/1
  - https://github.com/nodejs/node/issues/2833

- If dream with tests and code coverage, there is a bunch of repos that could use some love
  - https://github.com/ipfs/node-ipfs#roadmap

**Comment by diasdavid** [comment]
Aye! Check my new entry, arr :)

### LOG - Week of 21st of September

**Plan:**

- Have a interoperable version of libp2p in Go, to that the primitives and coverage become stronger so that then the remaining parts of IPFS can implemented with confidence (cli, bitswap, chunker, etc)

**Where you can help during this week:**

- Making SPDY become a duplex stream - https://github.com/indutny/spdy-transport/issues/14

- MOAR transports. Create other transports that swarm can use and tests to see them working at the same time
  - [libp2p-udp](https://github.com/diasdavid/node-libp2p-udp)
  - [libp2p-udt](https://github.com/diasdavid/node-libp2p-udt)
  - [libp2p-utp](https://github.com/diasdavid/node-libp2p-utp)

- If dream with tests and code coverage, there is a bunch of repos that could use some love
  - https://github.com/ipfs/node-ipfs#roadmap

**Comment by diasdavid** [comment]
### LOG - Week of December 21

**Update:**

- We have now IPFS Repo implemented in JS https://github.com/ipfs/js-ipfs-repo
- Bitswap work has started 
  - https://github.com/diasdavid/js-ipfs-bitswap
  - https://github.com/ipfs/specs/pull/53

**Next:**

- Make a IPLD Data Importing spec https://github.com/ipfs/js-ipfs/issues/41#issuecomment-164370167
- jsipfs init - https://github.com/ipfs/js-ipfs/issues/42
- jsipfs daemon

**Notes:**

 - For a specific libp2p log, check https://github.com/diasdavid/js-libp2p/issues/6

**Comment by diasdavid** [comment]
### LOG - 26 of January

#### Update:

- You can now do `npm i -g jsipfs` and use the already implemented commands, full list on project status https://github.com/ipfs/js-ipfs#project-status
- We have an MerkleDAG implementation that interops with go-ipfs MerkleDAG implementation (this means you can write and read blocks from the same repo and will work for both implementations)
- We have currently a WIP on browser testing, with only one obstacle from storing properly binary data in the browser
- Lot's of good stuff

#### Next:

- `files add` do the whole file/dir -> MerkleDAG thing, following the Data Importing spec https://github.com/ipfs/specs/pull/57

#### Where you can help?

In lots of places, check https://github.com/ipfs/js-ipfs/issues for descriptions, there is a huge amount of things that were unblocked once IPFS Repo and MerkleDAG were implemented and that can be done in parallel :) Once we get this all done, we will focus again on libp2p to get the networking stack ready for prime time!

Thank you!

**Comment by diasdavid** [comment]
# LOG - February 23 🚣

![](https://media.giphy.com/media/IehS5CoFfzQuQ/giphy.gif)

We've been pretty busy working on the JavaScript implementation of IPFS, there are more things that you can use now and all of them interop the Go implementation. Before starting with the formal update and next steps, I just wanted give a shout to everyone that has been submitting patches or contributing with features to js-ipfs lately, special thanks goes to:

- [@vijayee](https://github.com/vijayee/)
- [@dignifiedquire](https://github.com/dignifiedquire)
- [@xicombd](https://github.com/xicombd/)
- [@nginnever](https://github.com/nginnever)

## Update

- [x] IPFS MerkleDAG (DAGNode + DAGLink) and DAGService https://github.com/vijayee/ipfs-merkle-dag 
- [x] IPFS Blocks (block data structure and Block Service) https://github.com/ipfs/js-ipfs-blocks
- [x] IPFS unixfs https://github.com/ipfs/js-ipfs-unixfs 
- [x] IPFS Architecture Overview https://github.com/ipfs/js-ipfs#ipfs-core-implementation-architecture
- [x] Import files and directories into IPFS by converting them to their MerkleDAG representations with ipfs-data-importing https://github.com/ipfs/js-ipfs-data-importing
- [x] object API is feature complete 
- [x] all js-ipfs core code is tested and it behaves correctly in the browser
- [x] js-peer-id and js-peer-info now use node-forge library to be able to run on the browser and generate the public key from a private key. (before it was built to know how to generate key pairs, but since go-ipfs stores only the private key, we had to adapt the js code)
- [x] jsipfs daemon
- [x] complete update on all the feature threads on the js-ipfs repo, adding more information and clear action items https://github.com/ipfs/js-ipfs/issues
- [x] level up js-ipfs tests

**Extra** - @RichardLitt has been putting an incredible amount of work in documenting the HTTP-API of IPFS, giving us a clear path on how to implement those routes correctly. You can find this endeavour in the [http-api-spec repo](https://github.com/ipfs/http-api-spec)

## Goals

Our short-term goals for js-ipfs is to have complete feature parity with go-ipfs, with the exception of DHT (now known as Peer Routing and Record Store), what this means is that if this goal is obtained, we will be able to:

- swarm listen and connect between js-ipfs and go-ipfs nodes
- use bitswap
- run IPFS in the browser

This will mark a very significant milestone on the track of the js-ipfs development, enabling other developers to start build apps and services using the JavaScript libraries.

Peer Routing and Record Store are the stretch goals, since it would require go-ipfs to learn about IPLD and IPRS. 

## Good foundations

We've invested a significant portion of time in making tests from the beginning of the development of this project. Tests makes the codebase less brittle and help contributors feel confident that their patches and features don't break any other feature.

js-ipfs is divided at the macro level in ['ipfs-core', 'http-api' and 'cli'](https://github.com/ipfs/js-ipfs#project-structure). It all starts with core, it is where all the functionality that IPFS is composed of, is exposed through a standardized interface ([WIP](https://github.com/ipfs/specs/tree/master/api)), each feature that gets implemented must be [accompanied by tests and these must run in Node.js and the Browser](https://github.com/ipfs/js-ipfs/tree/master/tests/test-core). In addition, each time one of these features manipulares blobs of data that exit the memory of the program (for e.g., when they are written to disk), they must be validated and compared to the ones that are produced by go-ipfs, ensuring interoperability and parity ([example](https://github.com/ipfs/js-ipfs-data-importing/blob/master/tests/test-import.js))
 
Then, the http-api is built, which uses core and exposes its features through a RPC like HTTP API, specified by the [http-api-spec](https://github.com/ipfs/http-api-spec). [Tests are built](https://github.com/ipfs/js-ipfs/blob/master/tests/test-http-api/test-id.js) by testing the http-api directly and by using the [js-ipfs-api](https://github.com/ipfs/js-ipfs-api), a module that is up do date with go-ipfs implemented http-api.

Finally, comes the CLI, which should be tested by command, while an daemon is online and offline.

**Let's look at an example, the `jsipfs id` feature**

- core implementation https://github.com/ipfs/js-ipfs/blob/master/src/ipfs-core/index.js#L44-L66
- core tests https://github.com/ipfs/js-ipfs/blob/master/tests/test-core/test-id.js
- http api implementation https://github.com/ipfs/js-ipfs/blob/master/src/http-api/routes/id.js and https://github.com/ipfs/js-ipfs/blob/master/src/http-api/resources/id.js
- http-api tests https://github.com/ipfs/js-ipfs/blob/master/tests/test-http-api/test-id.js#L7-L24
- http-api tests with js-ipfs-api https://github.com/ipfs/js-ipfs/blob/master/tests/test-http-api/test-id.js#L28-L44
- cli implementation https://github.com/ipfs/js-ipfs/blob/master/src/cli/commands/id.js
- cli tests with js-ipfs daemon not running https://github.com/ipfs/js-ipfs/blob/master/tests/test-cli/test-id.js#L8-L34
- cli tests with js-ipfs daemon running https://github.com/ipfs/js-ipfs/blob/master/tests/test-cli/test-id.js#L36-L76

## Where can you help

- By completing one of the action items available on the 'help wanted' section https://github.com/ipfs/js-ipfs/issues?q=is%3Aissue+is%3Aopen+label%3A%22help+wanted%22
- Data Importing "export" feature https://github.com/ipfs/js-ipfs-data-importing/blob/master/src/index.js#L174-L176
- Improve code coverage - Help us reach 100% across every JavaScript repo!
- Review the last http-ipfs-api spec PR https://github.com/ipfs/http-api-spec/pulls

**Comment by diasdavid** [comment]
# LOG - April 5 📖

![](https://media.giphy.com/media/xT9DPn2xTR5mVs0mRO/giphy.gif)

**We’re close!™**. That is right, we are indeed pretty close to get a workable JavaScript implementation of IPFS that can be ran on the browser. I want to thank first, everyone that has been actively contributing to js-ipfs, with features, tests, documentation and experimentation! This participation is key to get the js-ipfs implementation out the door as soon as possible.

## Update

We are now able to use Web Sockets with SPDY stream muxing, giving us the change to use js-ipfs on the browser and exchange blocks with any IPFS Node that is available through a public IP. Big ^5 to @dignifiedquire who fixed pako, the JavaScript implementation of zlib, the last mile to make this work completely.

@ngginever has been leading the charge on the `files api` front, `jsipfs files get` is almost in a ready state to be merged, powered by pifs-data-importing pipeline.

@noffle finished `jsipfs init` now works and made it fully compatible with go-ipfs. While at it, some documentation for it was also created as a bonus.

Overall js-ipfs keeps getting more and more complete. @xicombd lead most of the charges on getting the http-api endpoints and cli commands implemented, using the same IPFS core codebase.

We’ve updated our [ROADMAP](https://github.com/ipfs/js-ipfs/blob/master/ROADMAP.md) to make it even more clear where we are and where he are headed.

## Goal

We’ve set up a goal of getting `js-ipfs` milestone 1 ready by the 25th of April. It is a ambitious and there might be some hiccups that delay this goal, but we feel confident as we are pretty close to be fulfil all the goals. Nevertheless, hitting milestone 1 is not just about getting something that works, it is about getting something that works and that people feel confident to use, that being said, one of the things we want to invest time in is on having good documentation of all the pieces that compose js-ipfs and test coverage.

## Where you can help

There are a lot of things where everyone can help, I’ve opened a issue for each of them with a task list, some focused on ‘features’, others focused in ‘modules’.

Here is the list of issues with what we are looking for to ship before milestone 1:

- https://github.com/ipfs/js-ipfs/issues/60
- https://github.com/ipfs/js-ipfs/issues/126
- https://github.com/ipfs/js-ipfs/issues/127
- https://github.com/ipfs/js-ipfs/issues/128
- https://github.com/ipfs/js-ipfs/issues/92
- https://github.com/ipfs/js-ipfs/issues/95
- https://github.com/ipfs/js-ipfs/issues/110
- https://github.com/ipfs/js-ipfs/issues/111
- https://github.com/ipfs/js-ipfs/issues/112
- https://github.com/ipfs/js-ipfs/issues/113
- https://github.com/ipfs/js-ipfs/issues/114
- https://github.com/ipfs/js-ipfs/issues/115
- https://github.com/ipfs/js-ipfs/issues/116
- https://github.com/ipfs/js-ipfs/issues/117
- https://github.com/ipfs/js-ipfs/issues/118
- https://github.com/ipfs/js-ipfs/issues/119
- https://github.com/ipfs/js-ipfs/issues/120
- https://github.com/ipfs/js-ipfs/issues/121
- https://github.com/ipfs/js-ipfs/issues/122
- https://github.com/ipfs/js-ipfs/issues/123
- https://github.com/ipfs/js-ipfs/issues/124
- https://github.com/ipfs/js-ipfs/issues/125

You will always be able to find in this repo, a collection of issues tagged with [`help wanted`](https://github.com/ipfs/js-ipfs/issues?q=is%3Aissue+is%3Aopen+label%3A%22help+wanted%22). This tag means that there is an action item that is currently not blocked by any other endeavour.

**Comment by diasdavid** [comment]
# LOG - April 5 🐳

Aye! This is a short captain.log entry to update everyone on the state of things. It is a short entry because we are really close of having `js-ipfs` ready and because of that, there is a set of very specific focus points we need to put our attention too. This entry enumerates those action items.

# What's left

**files API**

- [ ] get unixfs-engine fully operational in the browser. Currently there is a bug caused by changes on the readable-stream impl - https://github.com/ipfs/js-ipfs-unixfs-engine/issues/31. @dignifiedquire is handling this
- [ ] finish `add`, `get` and `cat`commands on `jsipfs cli. This includes through daemon (multipart messages) and directly calling core. @nginnever is tackling these. Follow on https://github.com/ipfs/js-ipfs/issues/60

**libp2p**

- [ ] swarm random bug https://github.com/diasdavid/js-libp2p-swarm/issues/37
- [ ] update js-multistream to follow the latest refinement on the spec https://github.com/jbenet/multistream/pull/7#discussion_r60112246 . Ensure it is compatible with go-multistream
- [ ] Ensure interop (this requires go-ipfs to follow all the multicodecs described https://github.com/ipfs/specs/pull/94)

**bitswap**

- [ ] 1st part: Use bitswap to fetch blocks (able to request and cancel with the message queue + state machine architecture discussed)
- [ ] 2nd part: Serve blocks

**module quality**

We are mostly done with our module quality improvements (code coverage, readme, etc), down to 5 modules, namely:
- [ ] js-libp2p-ipfs https://github.com/ipfs/js-ipfs/issues/125
- [ ] js-libp2p-tcp https://github.com/ipfs/js-ipfs/issues/123
- [ ] js-libp2p-swarm https://github.com/ipfs/js-ipfs/issues/122
- [ ] js-libp2p-unixfs-engine https://github.com/ipfs/js-ipfs/issues/112
- [ ] js-ipfs-api

@noffle is helping getting these done. If you want to help, jump in into one of those issues as contribute with one of the items that is left, feel also welcome to post questions :)

**Comment by diasdavid** [comment]
# LOG - May 10 🚤

Arr! This is the last mile Captain.log for js-ipfs milestone 1. It contains the list of things that are missing in order to js-ipfs usable!

# What's left

**files API**

- [ ] `add`, `get` and `cat` HTTP API endpoints. https://github.com/ipfs/js-ipfs/pull/205
- [ ] js-ipfs-api `get` feature implemented https://github.com/ipfs/js-ipfs-api/issues/263

**libp2p**

- [ ] update js-multistream to follow the latest refinement on the spec https://github.com/jbenet/multistream/pull/7#discussion_r60112246 . Ensure it is compatible with go-multistream
- [ ] support the spec defined multicodec in go-ipfs https://github.com/ipfs/go-ipfs/issues/1549
- [ ] js-libp2p-ipfs + go-libp2p interop tests

**bitswap**

- [ ] browser tests
- [ ] intensive tests (bitswap swarm between several browser and Node.js nodes)

**module quality**

We are mostly done with our module quality improvements (code coverage, readme, etc), down to 5 modules, namely:

- [ ] js-ipfs-api

**core api**

- [ ] converge both js-ipfs-api and js-ipfs interfaces so that modules are interchangeable
- [ ] core api spec https://github.com/ipfs/specs/issues/85

**Comment by diasdavid** [comment]
# LOG - June 14 ⛵

![](http://www.picgifs.com/graphics/b/boats/graphics-boats-897332.gif)

Aye mates! This log entry is full of exciting updates, writing it made me really happy. The IPFS community has worked pretty hard and there is a lot to talk about and a lot of people talking and checking IPFS, and that is amazing!

## What’s new in town

We now have two apps working on js-ipfs, in the browser, completely written in JavaScript! These are:

- Orbit
- Paperhub

#### Orbit

> http://orbit.libp2p.io/

Distributed Chat Application using CRDT

Code at: https://github.com/haadcode/orbit

#### Paperhub

> http://paperhub.libp2p.io/

Collaboration tool for distribution of Open Access Scientific material.

Code at: https://github.com/ipfs/paperhub

Both of these apps use [`orbit-db`](https://www.npmjs.com/package/orbit-db), a DB for P2P networks that uses a CRDT to synchronise the state between several peers. You can find [here] a fun demo and visualisation.

http://celebdil.benet.ai:8080/ipfs/Qmezm7g8mBpWyuPk6D84CNcfLKJwU6mpXuEN5GJZNkX3XK/

Huge thank you and lot’s of <3 <3 <3 to everyone that has been contributing so much to the project.

## News

The IPFS project was present at the Decentralised Web Summit, hosted by the Internet Archive. Expect a lot of blog posts from us soon, meanwhile, you can check the lighting talks at https://archive.org/details/DecentralizedWeb20160609pt1

## What’s next

There is still a lot of fun things to tackle, specially when it comes to interop between go-ipfs and js-ipfs, me and @whyrusleeping have been inspecting the wire protocols and found some differences in the implementations, most of them have been fixed, missing now only the [Identify](https://github.com/diasdavid/js-libp2p-swarm/pull/77) one. You can find the list below of things that are on the plan for this month and where you can provide most help too.


#### go and js interop

- libp2p-swarm
    - revamp the Identify protocol implementation https://github.com/diasdavid/js-libp2p-swarm/issues/78

#### libp2p Transports + Connection

- transports
    - [ ] upgrade the interface-transport spec https://github.com/diasdavid/interface-transport/pull/6
        - take more inspiration from the net package
        - add an option to timeout (destroy all the conns if not closed graciously)
        - Make each listener be instantiated separately (less magic 
- connection
    - [ ] upgrade the interface-connection to have
        - getObservedAddrs
        - getPeerInfo
            - Then, in swarm, overload it, so that it uses identify if needed ⍟
        - setPeerInfo
        - https://github.com/diasdavid/interface-connection/pull/4/files
        - create Connection class

#### libp2p UX

- libp2p
    - [ ] js-libp2p base, that libp2p-ipfs and libp2p-browser can Inherit, avoiding code duplication
    - [ ] define interface-libp2p

#### libp2p Features

- [ ] multicastDNS https://github.com/diasdavid/js-libp2p-mdns-discovery
    - enable it again
    - dependent on getting interop first (avoid weird scenarios) 
- [ ] add ping protocol again to js-libp2p (https://github.com/diasdavid/js-libp2p-ping)
- [ ] secio https://github.com/diasdavid/js-libp2p-swarm/pull/67

#### Testing

- [ ] Upgrade the tests scripts to be self contained . This will improve iteration speed by saving a lot of time on running all the tests every single time. Optimally, they even should be runnable with `node /test/test.js` (not needing mocha directly)
- [ ] 

#### CLI

- [ ] Improve CLI
    - [ ] detect incoherences with go-ipfs CLI

#### js-ipfs

- interface-ipfs-core
    - [ ] block API
    - [ ] id
    - [ ] version
    - [ ] pinning API
    - [ ] swarm API
    - [ ] DHT API
    - [ ] ping
- logging
    - [ ] be more diligent on adding debug logs 

#### Documentation

    - Better examples on how to use libp2p
    - Examples on how to use js-ipfs
        - https://github.com/ipfs/js-ipfs/issues/128
        - https://github.com/ipfs/js-ipfs/issues/127
    - libp2p spec upgrade

#### General bug fixing

- MerkleDAG
    - Fix https://github.com/vijayee/js-ipfs-merkle-dag/pull/40
- Orbit https://github.com/haadcode/orbit/issues/32

#### Peer endeavours

- ipfsd-ctl
    - Upgrade it to use bin-wrapper instead
    - https://github.com/ipfs/js-ipfsd-ctl/issues/78
- registry-mirror -> ip-npm
    - Upgrade API
    - Leverage the new UnixFS Sharding
- station
    - use the new ipfsd-ctl
- IPLD playground
- beaker integration
    - https://github.com/pfraze/beaker/issues/2

#### Stretch goals

- Interop testing automation https://github.com/ipfs/ipfs-interop
- bitswap benchmarking 
- js-ipfs compatible with iptb (testbed)
    - get Dockerfile
    - use iptb to ssh into the nodes

**Comment by diasdavid** [comment]
# LOG - August 13

![image](https://cloud.githubusercontent.com/assets/1211152/17645228/7b8b11f2-6197-11e6-9efc-c459f2cef599.png)

> Aye mates! Here is another exciting update on the js-ipfs project. This one contains: a new and extended Roadmap for the project; some really exciting feature updates; a description of where our attention is and a plug to our new project tracking process and as usual, a up to date list of items that you can tackle and contribute to js-ipfs!

# 🌬 Release the sails! We've found a (road)map

The IPFS community has spent some quality time in July together (with in person + remote attendance) at the [IPFS Workshop](https://github.com/ipfs/2016-Q3-Workshop) to come up with a plan for what are the new and exciting things coming up to the IPFS ecosystem. From those discussions, a new shiny roadmap emerged with very focused milestones, so that we can complete and make js-ipfs better incrementally, while keeping the focus in a subset of tasks at a time and ensure those are completed excellently.

You can find the [Full Roadmap](https://github.com/ipfs/js-ipfs/blob/master/ROADMAP.md) at this project repo, however, here is a tl;dr; for your reading pleasure:

- [✔️ ~~Milestone 1 - js-ipfs on the browser~~](https://github.com/ipfs/js-ipfs/blob/master/ROADMAP.md#milestone-1---js-ipfs-on-the-browser)
- [Milestone 2 - Quality and correctness of the JavaScript Implementation](https://github.com/ipfs/js-ipfs/blob/master/ROADMAP.md#milestone-2---quality-and-correctness-of-the-javascript-implementation)
- [Milestone 3 - Enter IPLD - Enable js-ipfs to handle both merkledag protobufs and ipld cbor objects](https://github.com/ipfs/js-ipfs/blob/master/ROADMAP.md#milestone-3---enter-ipld---enable-js-ipfs-to-handle-both-merkledag-protobufs-and-ipld-cbor-objects)
- [Milestone 4 - Documentation and Developer Experience](https://github.com/ipfs/js-ipfs/blob/master/ROADMAP.md#milestone-4---documentation-and-developer-experience)
- [Milestone 5 - 1st implementation Peer Routing + Content Routing (the go-ipfs DHT)](https://github.com/ipfs/js-ipfs/blob/master/ROADMAP.md#milestone-5---1st-implementation-peer-routing--content-routing-the-go-ipfs-dht)
- [Milestone 6 - The last miles for complete feature parity with go-ipfs](https://github.com/ipfs/js-ipfs/blob/master/ROADMAP.md#milestone-6---the-last-miles-for-complete-feature-parity-with-go-ipfs)

# 🌟 Moar exciting news!

- Moving to pull-streams - The 🚲🏠 was strong in this one, but we've finally made a call on the Control Flow library to use in order to solve the hiccups we have had in the past with Node.js Streams. This discussion took a lot of input from several people familiar with the js-ipfs project and the contenters for the Control Flow Library. You can read the whole thread at: https://github.com/ipfs/js-ipfs/issues/362
- Automated Changelogs - We are getting better changelogs on release that are automatically generated. We've finalized our commit message convention, [here](https://github.com/ipfs/community/blob/master/js-commit-guidelines.md), so that they easily parsable and human understandable. The issue to track the development of the tooling required to add this feature to our release process is [here](https://github.com/dignifiedquire/aegir/issues/30#issuecomment-236685903)
- On top of the changelog, we also created a channel to announce and propose changes to the js-ipfs interface. If you want to be notified about these, please subscribe to this [issue](https://github.com/ipfs/interface-ipfs-core/issues/55)
- On the topic of js-ipfs interface, we've updated the files.add API to become more Developer friendly. Read more in this [issue](https://github.com/ipfs/interface-ipfs-core/issues/29)
- Our js-ipfs got an upgrade. We moved from ronin to yargs, a more supported and used CLI toolkit that enables us to have menus for the js-ipfs CLI just like go-ipfs has.
- We've started optimizing js-ipfs and talking about future optimizations, to make the library smaller and more performant - Full thread [here](https://github.com/ipfs/js-ipfs/issues/398)

# ⚡️ Current focus

Our current focus is now identified by the Milestone where the team is currently hacking on (Milestone 2). In addition to the [ROADMAP.md](https://github.com/ipfs/js-ipfs/blob/master/ROADMAP.md), we've created a [waffle board for the js-ipfs project](https://waffle.io/ipfs/js-ipfs), so that we (including you) can easily identify easily which issues are related with a certain Milestone.

Important to note that this new [Project Management Process is still being finalised, you can follow it on the PM repo.](https://github.com/ipfs/pm/pull/131)

# 💖 Where you can help

As always, issues that are tagged with 'help wanted' are open for contribution and that label means that they are unblocked, plus they are typically tagged with a difficulty level (let us know if we missed one). 

Check our waffle board for a complete list across projects: https://waffle.io/ipfs/js-ipfs

If you are new to the project and want to get a better feeling of how are things, an always good way is to join our Monday hangouts. You can find all the details on the [PM repo](https://github.com/ipfs/pm).

# 📰 Other things that might be of your interest

- `ipfsd-ctl` is being revamped to use bin-wrapper and with it, give a facelift to the documentation. A lot of developers (ourselves included) use this module to spawn go-ipfs instances, it is super handy for testing and electron apps and now it is getting better. Thank you @JGAntunes for leading that!
- Now libp2p is a official standalone project with its own organization, you can find all of its modules on the [libp2p organization](https://github.com/libp2p)
- Our self described formats have been promoted and now are standalone endeavour, in the process of standardization. See [multiformats](https://github.com/multiformats) for more info.
- **PubSub Initiative** - We've kicked off officially the endeavour to bring a PubSub system for IPFS. A lot of people in the IPFS community have joined to make this happen, to watch or join this effort, see: https://github.com/libp2p/pubsub
- **IPLD is coming, closer than ever** - A new proposal for Content Identifiers (CID) has been made and with that a migration plan for IPFS to support both MerkleDAG and IPLD object (+ any other data structure), this is super exciting, find it all at: https://github.com/ipfs/specs/issues/130 
- go-ipfs 0.4.3 Release Candidate is out, read all about it on its Captain.log - https://github.com/ipfs/go-ipfs/issues/2247#issuecomment-238675394

# Thank you!

Thank you for reading this entry of the captain.log and a bigger thank you to everyone that has been contributing, either through code, bug fixing, bug reporting, ideas, reviews and so on, you are awesome!!!

**Comment by diasdavid** [comment]
Hi everyone, this is your Captain speaking. I'm aware I haven't shared an update here in a while (mainly due we opened new channels like the waffle board, the roadmap, awesome endeavors and so on), but I didn't want to miss the opportunity to capture your attention to Issue https://github.com/ipfs/js-ipfs/issues/795#issue-214633515, the 0.23 release.

v0.23.0 brings a lot of exciting new things, including interop with go-ipfs which many of you waited (for so long!). Make sure to check it out :) Thank you.

**Comment by diasdavid** [comment]
I'm officially closing this edition of the Captain.log. It was fun times to write this update to you all but overtime it become a too large time sink and with very little impact given that we have now:

- Weekly Calls https://github.com/ipfs/js-ipfs/blob/master/MGMT.md
- Project Tracking https://github.com/ipfs/js-ipfs/blob/master/MGMT.md
- Release Issues - #1320 

See you in the other channels!
